### PR TITLE
Generalized `rock::ZeroInitKernelOp` ->  `rock::InitKernelOp`

### PR DESCRIPTION
**Pull request by ravil-mobile**
* Added an optional attr. which can specify init. value
* The buffer will be filled with zeros if the init. value is not provided

Closes ROCm/rocMLIR-internal#1324

### CODE REVIEW

**Comment by ravil-mobile** [comment]
Hi @krzysz00 

I tried the following

```cpp
    if (initValueAttr) {
      auto tupedAttr = initValueAttr.value().cast<TypedAttr>();
      initOp = b.create<ConstantOp>(loc, storeType, tupedAttr);
    }
    else {
      initOp = createZeroConstantOp(b, loc, storeType);
    }
```

This results in the following IR

```mlir
  %0 = "arith.constant"() <{value = 0.000000e+00 : f32}> : () -> vector<4xf32>
```

But we need

```mlir
  %0 = "arith.constant"() <{0.000000e+00 : f32}> : () -> vector<4xf32>
```

The following also doesn't work

```cpp
      if (auto floatInitValueAttr = initValueAttr->cast<FloatAttr>()) {
        auto a = floatInitValueAttr.getValue();
        initOp = b.create<ConstantOp>(loc, storeType, floatInitValueAttr.getValue());
      } else if (auto intInitValueAttr =
                     initValueAttr->cast<IntegerAttr>()) {
        initOp = b.create<ConstantOp>(loc, storeType, intInitValueAttr.getValue());
      } else {
        return failure();
      }
```

because `.getValue()` returns either APFloat or APInt. And there is not such a constructor for arith::ConstantOp

**Comment by krzysz00** [comment]
@ravil-mobile For that usecase, you'll probably want `vector::SplatOp`.

That or someone'll finally make the "make a constant of this scalar, broadcasting into a vector is needed" utility that's been written 20 times into an actual utility function on `Arith`. There's a decent change I'll do it early next week, but that's upstream changes anyway, so not immediately relevant to you.

File: mlir/include/mlir/Dialect/Rock/IR/RockOps.td [hunk 1]
```diff
@@ -245,18 +245,19 @@ def Rock_AttentionOp :
...
                   OptionalAttr<I32Attr>:$blockSize,
                   OptionalAttr<I32Attr>:$gridSize,
                   OptionalAttr<IndexAttr>:$elemsPerThread)>,
     Results<(outs Optional<AnyTensor>:$result)> {
-  let summary = "Zero-initialize a buffer";
+  let summary = "initialize a buffer";
 
   let description = [{
-    Zero-initialize a given buffer.
+    initialize a given buffer.
```
Comment (krzysz00) [inline mlir/include/mlir/Dialect/Rock/IR/RockOps.td]: Nit: capitalization

File: mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp [hunk 2]
```diff
@@ -267,41 +269,58 @@ static Value makeGpuAllocContaining(OpBuilder &b, Value v) {
...
     bool needs64BitIdx = is4GBMemoryType(buffer.getType());
-    Value zeroOp = createZeroConstantOp(b, loc, storeType);
+
+    Type elementType = mlir::getElementTypeOrSelf(storeType);
+    Value initOp;
+    auto initValueAttr = op.getInitValueAttr();
+    if (initValueAttr) {
+      if (auto floatInitValueAttr = initValueAttr.value().cast<FloatAttr>()) {
+        auto initValue = floatInitValueAttr.getValue().convertToFloat();
+        initOp =
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp:294]: Check if this works with vectors - I don't remember if it does

File: mlir/lib/Dialect/Rock/utility/builderUtils.cpp [hunk 3]
```diff
@@ -21,6 +21,19 @@ using mlir::arith::ConstantOp;
 
 namespace mlir {
 namespace rock {
+FloatAttr getConstantScalarAttr(OpBuilder &b, Type elementType, float value) {
+  APFloat apValue(value);
+  auto constValue = b.getFloatAttr(elementType, apValue);
+  return constValue;
+}
+
+IntegerAttr getConstantScalarAttr(OpBuilder &b, Type elementType,
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/utility/builderUtils.cpp]: Unused, not declared in headers?

**Review by krzysz00 — APPROVED** [review]
Overall, looks fine, just minor nitpicks

File: mlir/include/mlir/Dialect/Rock/IR/RockOps.td (hunk 1 above)
Comment (ravil-mobile) [reply mlir/include/mlir/Dialect/Rock/IR/RockOps.td]: Done

File: mlir/lib/Dialect/Rock/utility/builderUtils.cpp (hunk 3 above)
Comment (ravil-mobile) [reply mlir/lib/Dialect/Rock/utility/builderUtils.cpp]: Thanks for noticing! I completely forgot to delete these functions.
Done

File: mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp (hunk 2 above)
Comment (ravil-mobile) [reply mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp:294]: Hi @krzysz00, I tried `vector::SplatOp`. I noticed that the operation requires `type` and `arith::ConstantOp`. I tried to pass `TypedAttr` or `APFloat/APInt` to the constructor, but it didn't work

Comment (krzysz00) [reply mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp:294]: Ah. Well then we'll clean this up Later (tm), it's not a big deal

**Review by krzysz00 — APPROVED** [review]
Approved!

**Pull request merged**